### PR TITLE
Add option for Azure cross subscription backups

### LINKED ISSUE #1646: Azure: cross subscription backups
**What steps did you take and what happened:**
Background: We have multiple AKS clusters in different subscriptions ( same tenant ). We are aiming for a single storage account were we can store backups from all clusters, in different buckets though.
- The storage account resides in a subscription by it self.
- Service principals are created in each subscription with a cluster and given contributor to the backup storage account and resource group.

The problem is we cant tell Velero that the resource group and storage account resides in another subscription than the service principal, even though it have access to the resource.

**The output of the following commands will help us better understand what's going on**:
(Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

Tried to specify the whole path to the resource group:
`time="2019-07-11T10:29:34Z" level=info msg="Checking that all backup storage locations are valid" logSource="pkg/cmd/server/server.go:431"
An error occurred: some backup storage locations are invalid: error getting backup store for location "default": rpc error: code = Unknown desc = storage.AccountsClient#ListKeys: Invalid input: autorest/validation: validation failed: parameter=resourceGroupName constraint=Pattern value="/subscriptions/REDACTED/resourceGroups/REDACTED" details: value doesn't match pattern ^[-\w\._\(\)]+$`

And when i just enter the name of the group:
`time="2019-07-11T10:15:44Z" level=info msg="Checking that all backup storage locations are valid" logSource="pkg/cmd/server/server.go:431"
An error occurred: some backup storage locations are invalid: error getting backup store for location "default": rpc error: code = Unknown desc = storage.AccountsClient#ListKeys: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceGroupNotFound" Message="Resource group 'REDACTED' could not be found."`

**Environment:**

- Velero version (use `velero version`): 1.0.0
- Kubernetes version (use `kubectl version`): 1.13.5
- Cloud provider or hardware configuration: AKS

### PR DESCRIPTION
**Pull request by boxcee**
Closes #1646

### CODE REVIEW

**Comment by skriss** [comment]
Thanks for this @boxcee! I'll review and do some testing this week.

File: pkg/cloudprovider/azure/object_store.go [hunk 1]
```diff
@@ -146,22 +147,31 @@ func getStorageAccountKey(config map[string]string) (string, error) {
 		return "", errors.Wrap(err, "unable to get all required environment variables")
 	}
 
-	// 2. we need config["resourceGroup"], config["storageAccount"]
+	// 2. check whether a different subscription ID was set for backups in config["subscriptionId"]
+	var subscriptionId string
```
Comment (skriss) [inline pkg/cloudprovider/azure/object_store.go]: I'd probably simplify this section to something like:

```go
subscriptionID := envVars[subscriptionIDEnvVar]
if val := config[subscriptionIdConfigKey]; val != "" {
	subscriptionID = val
}
```

Comment (boxcee) [reply pkg/cloudprovider/azure/object_store.go]: Sure, thanks. Just digging into the Go stuff.

**Comment by boxcee** [comment]
@skriss For testing, is there anything I should document? We have an Azure account and I will test this today.

**Comment by skriss** [comment]
@boxcee just some quick notes on exactly what scenario/steps you tested would be great - don't need anything long, just enough to get the idea. thanks!

**Comment by boxcee** [comment]
Soooo, I did test this and it does work. **However**, I am not really sure what value it adds as the snapshots cannot be stored in another subscription. That is at least as far as I got.

I wanted to test this by creating a cluster, creating a backup, deleting the cluster, creating another cluster and then restore the backup to that cluster. Turns out, when I deleted the cluster I also deleted all the snapshots. When I then tried to restore, the snapshots could not be found.

**Comment by jweldonca** [comment]
hey @boxcee, how do you test this succesfully? 

When I build you're branch (boxcee:allow-cross-subscription-backups) and then run velero install

I get the following error 
`An error occurred: some backup storage locations are invalid: error getting backup store for location "default": rpc error: code = Unknown desc = config has invalid keys [subscriptionId]; valid keys are [resourceGroup storageAccount bucket prefix]
`

This is the command I've ran
`velero install --provider azure --bucket $BLOB_CONTAINER --secret-file ./credentials-velero --backup-location-config resourceGroup=$rg,storageAccount=$storage,subscriptionId=$subId --snapshot-location-config apiTimeout=10000,resourceGroup=$rg,subscriptionId=$subId`

**Comment by boxcee** [comment]
@jweldonca I assume this error shows up in the controller?

By default the `Makefile` creates a new container when run with `all-containers`. I changed the repository to point to our company repository and docker pushed the images there.
Then I also modified the deployment (after installing with velero install) to point to the company repo.

Your problem is basically that you have built a new velero binary, but the images being deployed are still the `master` images.

**Comment by boxcee** [comment]
@skriss Can confirm it does work cross-subscription now!

1. Created a cluster in subscription A and configured it to create snapshots and backups in subscription B.
2. Used the example `nginx-app` to create a backup.
3. Deleted the cluster in subscription A.
4. Re-created a cluster in subscription A and configured it as before to use subscription B for snapshots and backups.
5. Then I ran the restore command from the example.
6. Success!

**Comment by skriss** [comment]
👍 I didn't realize that cross-subscription volume snapshots were even possible - sounds great! I'll take another look at the code shortly.

**Pull request merged**